### ISSUE TITLE
Snakes Losing Points On Win

### ISSUE THREAD

**Issue by GriffinLedingham** [battlesnakeio/roadmap#233]
Snakes are periodically losing points following a win on play.battlesnake.io

This has been occurring for myself (OpieToo), as well as Arbok, and various other snakes. It seems to be mostly when highly ranked on the leaderboard, and when games run short/are against lower ELO opponents.

Examples of some games this occurred in:

https://play.battlesnake.io/games/2bef77fc-a52e-4ed5-a428-14ffd7279404/

https://play.battlesnake.io/games/01fca5e6-166f-49e6-99a4-a21365d4c4e5/

https://play.battlesnake.io/games/88fadc4e-32b7-4b1b-a498-29e791112a45/

https://play.battlesnake.io/games/31f13080-9825-4be5-9b35-23db556b5508/

https://play.battlesnake.io/games/d6861700-b2f1-4ee0-aa84-91310f609798/

https://play.battlesnake.io/games/4fe7d7c5-bfd5-4680-b5c3-08c218cf0609/

https://play.battlesnake.io/games/ff297e60-59f0-4ee0-9bc6-4cec22fe0494/

Submitted by Weslord on Slack:

https://play.battlesnake.io/games/1da6bf92-52c5-471f-bd5f-593c0d4cfe10/

https://play.battlesnake.io/games/be596419-1be3-40bd-8cad-37036154d0a0/

**Comment by codeallthethingz** [comment]
@GriffinLedingham we have investigated this and we're a little stumped.  
We'd be super happy if you could find where this issue is as we're currently swamped working on getting ready for the tournament.

**Comment by codeallthethingz** [comment]
Actually, just realized we've taken play private so that's currently closed source.

**Comment by GriffinLedingham** [comment]
@codeallthethingz I've just read a fair bit about the technicalities of TrueSkill, and I'm wondering if this is intended behaviour after all. If the change in sigma is greater than the change in mu after a game, I'm wondering if the deviation can cause an unpredictable "rating loss" as it's being referred to here. 

I think that some of the trouble here comes from TrueSkill being more intended as a behind-the-scenes matchmaking rating, versus something like ELO that behaves in a more predictable way to onlookers. ELO players can expect to see a rating increase after a win, with a worst case of +0, while ratings like TrueSkill and Glicko work on deviation and do weirder things to better matchmake people.

A ladder I've done a lot of rating work with is [Pokemon Showdown's](https://play.pokemonshowdown.com/ladder) which displays a number of different ratings per user on the ladder, but ultimately ranks players by the predictable ELO score. They use Glicko-1 and their own modified GXE rating to matchmake players, but as both of them are based on deviation values, ELO remains kings for ranking player in a ladder.

Something like ELO (modified for multiplayer games) may be worth looking into for the future of the leaderboard, while maintaining TrueSkill as the actual matchmaker as it does a better job placing players together in many cases.

**Comment by bvanvugt** [comment]
@GriffinLedingham thanks for your research and work here. Overall we're not super happy with how TrueSkill is turning out, but it was a quick win to get something live. We have big plans, but unsure yet if we'll get around to it before this year's event.

From our perspective the current Leaderboard acts more as venue for training and testing your snake, and less about ranking or standing (albeit, it's weirdly named given that purpose).

**Comment by cbinners** [comment]
I implemented trueskill on snakedown last year, and it is pretty paramount that games are not constructed from a random set of snakes. In order for the ranking system to function properly, the game must be constructed from a set of snakes that produce a "good" match, according to trueskill. In practice, this just means choosing snakes that have close rankings.

I'm not sure how the matchmaking is currently implemented, but if this is not done, I'm fairly certain implemented less random matches will improve both game quality and ranking accuracy.

Something to note is that the negative scores become much more frequent once the leaderboard grows. As the gap of snake scores grows, the system will _very_ quickly start to produce poor quality matches, which leads to negative scores, etc, for wins.

I'm happy to discuss with whomever feels like taking this on.

For reference, what I did on snakedown was choose a pivot point _P_ in the leaderboard, and sample _N_ snakes from a range of `[P-N, P+N]` in rank.
